### ISSUE TITLE
Unable to parse email KeyError: b'RFC822'

### ISSUE THREAD

**Issue by stefankahn** [domainaware/parsedmarc#45]
Hi. I have recently stood up parsedmarc using the install instructions on https://domainaware.github.io/parsedmarc/index.html
It works well using the cli to parse report from a local directory

running version 5.2.1 with Davmail

   DEBUG:__init__.py:905:Connecting to IMAP over plain text
   DEBUG:__init__.py:976:Found 2 messages in IMAP folder INBOX
   DEBUG:__init__.py:982:Processing message 1 of 2: UID 10
Traceback (most recent call last):
  File "/usr/local/bin/parsedmarc", line 11, in <module>
    sys.exit(_main())
  File "/usr/local/lib/python3.5/dist-packages/parsedmarc/cli.py", line 343, in _main
    strip_attachment_payloads=sa
  File "/usr/local/lib/python3.5/dist-packages/parsedmarc/__init__.py", line 987, in get_dmarc_reports_from_inbox
    ["RFC822"])[message_uid][b"RFC822"]
KeyError: b'RFC822'

2019-01-17 08:50:38,644 INFO  [ImapConnection-34854] davmail.connection  - LOGON - 127.0.0.1:34854 xxx@xx.xx.xx
2019-01-17 08:50:38,644 DEBUG [ImapConnection-34854] davmail  - > MIKK1 OK Authenticated
2019-01-17 08:50:38,644 DEBUG [ImapConnection-34854] davmail  - < MIKK2 LIST "" "Archive"
2019-01-17 08:50:38,692 DEBUG [ImapConnection-34854] davmail  - > * LIST (\HasChildren) "/" "Archive"
2019-01-17 08:50:38,693 DEBUG [ImapConnection-34854] davmail  - > MIKK2 OK LIST completed
2019-01-17 08:50:38,734 DEBUG [ImapConnection-34854] davmail  - < MIKK3 LIST "" "Archive/Aggregate"
2019-01-17 08:50:38,812 DEBUG [ImapConnection-34854] davmail  - > * LIST (\HasNoChildren) "/" "Archive/Aggregate"
2019-01-17 08:50:38,812 DEBUG [ImapConnection-34854] davmail  - > MIKK3 OK LIST completed
2019-01-17 08:50:38,854 DEBUG [ImapConnection-34854] davmail  - < MIKK4 LIST "" "Archive/Aggregate"
2019-01-17 08:50:38,933 DEBUG [ImapConnection-34854] davmail  - > * LIST (\HasNoChildren) "/" "Archive/Aggregate"
2019-01-17 08:50:38,933 DEBUG [ImapConnection-34854] davmail  - > MIKK4 OK LIST completed
2019-01-17 08:50:38,974 DEBUG [ImapConnection-34854] davmail  - < MIKK5 LIST "" "Archive/Forensic"
2019-01-17 08:50:39,054 DEBUG [ImapConnection-34854] davmail  - > * LIST (\HasNoChildren) "/" "Archive/Forensic"
2019-01-17 08:50:39,054 DEBUG [ImapConnection-34854] davmail  - > MIKK5 OK LIST completed
2019-01-17 08:50:39,094 DEBUG [ImapConnection-34854] davmail  - < MIKK6 LIST "" "Archive/Invalid"
2019-01-17 08:50:39,169 DEBUG [ImapConnection-34854] davmail  - > * LIST (\HasNoChildren) "/" "Archive/Invalid"
2019-01-17 08:50:39,169 DEBUG [ImapConnection-34854] davmail  - > MIKK6 OK LIST completed
2019-01-17 08:50:39,210 DEBUG [ImapConnection-34854] davmail  - < MIKK7 SELECT "INBOX"
2019-01-17 08:50:39,248 DEBUG [ImapConnection-34854] davmail.exchange.ExchangeSession  - Folder INBOX - Search items current count: 3 fetchCount: 500 highest uid: 12 lowest uid: 10
2019-01-17 08:50:39,248 DEBUG [ImapConnection-34854] davmail.exchange.ExchangeSession  - Message IMAP uid: 10 uid: AAAAAMJFXiqcxYZLn8LeyLhsBJ8BAOkEKOL0zPtKlch7lA7QutsAAAAADSgAAA== ItemId: AAMkAGZlODU3MGNkLThhNjEtNDZmOS05MjIxLTFkMjZjYzE5YjVhZgBGAAAAAADCRV4qnMWGS5/C3si4bASfBwDpBCji9Mz7SpXIe5QO0LrbAAAAAAEMAADpBCji9Mz7SpXIe5QO0LrbAAAAAA0oAAA= ChangeKey: CQAAABYAAADpBCji9Mz7SpXIe5QO0LrbAAAAABDQ
2019-01-17 08:50:39,248 DEBUG [ImapConnection-34854] davmail.exchange.ExchangeSession  - Message IMAP uid: 11 uid: AAAAAMJFXiqcxYZLn8LeyLhsBJ8BAOkEKOL0zPtKlch7lA7QutsAAAAADSkAAA== ItemId: AAMkAGZlODU3MGNkLThhNjEtNDZmOS05MjIxLTFkMjZjYzE5YjVhZgBGAAAAAADCRV4qnMWGS5/C3si4bASfBwDpBCji9Mz7SpXIe5QO0LrbAAAAAAEMAADpBCji9Mz7SpXIe5QO0LrbAAAAAA0pAAA= ChangeKey: CQAAABYAAADpBCji9Mz7SpXIe5QO0LrbAAAAABDR
2019-01-17 08:50:39,248 DEBUG [ImapConnection-34854] davmail.exchange.ExchangeSession  - Message IMAP uid: 12 uid: AAAAAMJFXiqcxYZLn8LeyLhsBJ8BAOkEKOL0zPtKlch7lA7QutsAAAAADSoAAA== ItemId: AAMkAGZlODU3MGNkLThhNjEtNDZmOS05MjIxLTFkMjZjYzE5YjVhZgBGAAAAAADCRV4qnMWGS5/C3si4bASfBwDpBCji9Mz7SpXIe5QO0LrbAAAAAAEMAADpBCji9Mz7SpXIe5QO0LrbAAAAAA0qAAA= ChangeKey: CQAAABYAAADpBCji9Mz7SpXIe5QO0LrbAAAAABDS
2019-01-17 08:50:39,248 DEBUG [ImapConnection-34854] davmail  - > * 3 EXISTS
2019-01-17 08:50:39,248 DEBUG [ImapConnection-34854] davmail  - > * 3 RECENT
2019-01-17 08:50:39,248 DEBUG [ImapConnection-34854] davmail  - > * OK [UIDVALIDITY 1]
2019-01-17 08:50:39,248 DEBUG [ImapConnection-34854] davmail  - > * OK [UIDNEXT 13]
2019-01-17 08:50:39,248 DEBUG [ImapConnection-34854] davmail  - > * FLAGS (\Answered \Deleted \Draft \Flagged \Seen $Forwarded Junk)
2019-01-17 08:50:39,248 DEBUG [ImapConnection-34854] davmail  - > * OK [PERMANENTFLAGS (\Answered \Deleted \Draft \Flagged \Seen $Forwarded Junk \*)]
2019-01-17 08:50:39,248 DEBUG [ImapConnection-34854] davmail  - > MIKK7 OK [READ-WRITE] SELECT completed
2019-01-17 08:50:39,334 DEBUG [ImapConnection-34854] davmail  - < MIKK8 UID SEARCH ALL
2019-01-17 08:50:39,378 DEBUG [ImapConnection-34854] davmail.exchange.ExchangeSession  - Folder INBOX - Search items current count: 3 fetchCount: 500 highest uid: 12 lowest uid: 10
2019-01-17 08:50:39,378 DEBUG [ImapConnection-34854] davmail.exchange.ExchangeSession  - Message IMAP uid: 10 uid: AAAAAMJFXiqcxYZLn8LeyLhsBJ8BAOkEKOL0zPtKlch7lA7QutsAAAAADSgAAA== ItemId: AAMkAGZlODU3MGNkLThhNjEtNDZmOS05MjIxLTFkMjZjYzE5YjVhZgBGAAAAAADCRV4qnMWGS5/C3si4bASfBwDpBCji9Mz7SpXIe5QO0LrbAAAAAAEMAADpBCji9Mz7SpXIe5QO0LrbAAAAAA0oAAA= ChangeKey: CQAAABYAAADpBCji9Mz7SpXIe5QO0LrbAAAAABDQ
2019-01-17 08:50:39,378 DEBUG [ImapConnection-34854] davmail.exchange.ExchangeSession  - Message IMAP uid: 11 uid: AAAAAMJFXiqcxYZLn8LeyLhsBJ8BAOkEKOL0zPtKlch7lA7QutsAAAAADSkAAA== ItemId: AAMkAGZlODU3MGNkLThhNjEtNDZmOS05MjIxLTFkMjZjYzE5YjVhZgBGAAAAAADCRV4qnMWGS5/C3si4bASfBwDpBCji9Mz7SpXIe5QO0LrbAAAAAAEMAADpBCji9Mz7SpXIe5QO0LrbAAAAAA0pAAA= ChangeKey: CQAAABYAAADpBCji9Mz7SpXIe5QO0LrbAAAAABDR
2019-01-17 08:50:39,378 DEBUG [ImapConnection-34854] davmail.exchange.ExchangeSession  - Message IMAP uid: 12 uid: AAAAAMJFXiqcxYZLn8LeyLhsBJ8BAOkEKOL0zPtKlch7lA7QutsAAAAADSoAAA== ItemId: AAMkAGZlODU3MGNkLThhNjEtNDZmOS05MjIxLTFkMjZjYzE5YjVhZgBGAAAAAADCRV4qnMWGS5/C3si4bASfBwDpBCji9Mz7SpXIe5QO0LrbAAAAAAEMAADpBCji9Mz7SpXIe5QO0LrbAAAAAA0qAAA= ChangeKey: CQAAABYAAADpBCji9Mz7SpXIe5QO0LrbAAAAABDS
2019-01-17 08:50:39,378 DEBUG [ImapConnection-34854] davmail  - > * SEARCH 10 11 12
2019-01-17 08:50:39,378 DEBUG [ImapConnection-34854] davmail  - > MIKK8 OK SEARCH completed
2019-01-17 08:50:39,418 DEBUG [ImapConnection-34854] davmail  - < MIKK9 UID FETCH 10 (RFC822)
2019-01-17 08:50:39,429 DEBUG [ImapConnection-34854] davmail.imap.ImapConnection  - * 1 FETCH (UID 10
2019-01-17 08:50:39,449 DEBUG [ImapConnection-34854] davmail.exchange.ExchangeSession  - Downloaded full message content for IMAP UID 10 (12048 bytes)
2019-01-17 08:50:39,449 DEBUG [ImapConnection-34854] davmail  - >  BODY[null] {12048}
2019-01-17 08:50:39,449 DEBUG [ImapConnection-34854] davmail  - > )
2019-01-17 08:50:39,449 DEBUG [ImapConnection-34854] davmail  - > MIKK9 OK UID FETCH completed
2019-01-17 08:50:39,539 INFO  [ImapConnection-34854] davmail.connection  - DISCONNECT - 127.0.0.1:34854

**Comment by seanthegeek** [comment]
I just pushed a change that will output the debugging info I need to fix this issue. Please use this command to upgrade your installation:

```bash
sudo -H pip3 install -U git+https://github.com/domainaware/parsedmarc
```

Then run `parsedmarc` again, and reply here with the output.

I'll use that output to push out another change with the fix.

Thanks!

**Comment by stefankahn** [comment]
Upgraded my install and re-ran.  The output looks the same

parsedmarc --save-aggregate --save-forensic --imap-no-ssl -H 127.0.0.1 --imap-port 1143 -u xx@xx.xx.xx -p xxxxxxx --debug 
   DEBUG:__init__.py:905:Connecting to IMAP over plain text
   DEBUG:__init__.py:976:Found 4 messages in IMAP folder INBOX
   DEBUG:__init__.py:982:Processing message 1 of 4: UID 10
Traceback (most recent call last):
  File "/usr/local/bin/parsedmarc", line 9, in <module>
    load_entry_point('parsedmarc==5.2.1', 'console_scripts', 'parsedmarc')()
  File "/usr/local/lib/python3.5/dist-packages/parsedmarc/cli.py", line 343, in _main
    strip_attachment_payloads=sa
  File "/usr/local/lib/python3.5/dist-packages/parsedmarc/__init__.py", line 987, in get_dmarc_reports_from_inbox
    ["RFC822"])[message_uid][b"RFC822"]
KeyError: b'RFC822'



Davmain log:

2019-01-18 09:26:47,494 DEBUG [ImapConnection-37118] davmail.exchange.ExchangeSession  - Message IMAP uid: 14 uid: AAAAAMJFXiqcxYZLn8LeyLhsBJ8BAOkEKOL0zPtKlch7lA7QutsAAAAADSwAAA== ItemId: AAMkAGZlODU3MGNkLThhNjEtNDZmOS05MjIxLTFkMjZjYzE5YjVhZgBGAAAAAADCRV4qnMWGS5/C3si4bASfBwDpBCji9Mz7SpXIe5QO0LrbAAAAAAEMAADpBCji9Mz7SpXIe5QO0LrbAAAAAA0sAAA= ChangeKey: CQAAABYAAADpBCji9Mz7SpXIe5QO0LrbAAAAABDW
2019-01-18 09:26:47,494 DEBUG [ImapConnection-37118] davmail  - > * 4 EXISTS
2019-01-18 09:26:47,495 DEBUG [ImapConnection-37118] davmail  - > * 2 RECENT
2019-01-18 09:26:47,495 DEBUG [ImapConnection-37118] davmail  - > * OK [UIDVALIDITY 1]
2019-01-18 09:26:47,495 DEBUG [ImapConnection-37118] davmail  - > * OK [UIDNEXT 15]
2019-01-18 09:26:47,495 DEBUG [ImapConnection-37118] davmail  - > * FLAGS (\Answered \Deleted \Draft \Flagged \Seen $Forwarded Junk)
2019-01-18 09:26:47,495 DEBUG [ImapConnection-37118] davmail  - > * OK [PERMANENTFLAGS (\Answered \Deleted \Draft \Flagged \Seen $Forwarded Junk \*)]
2019-01-18 09:26:47,495 DEBUG [ImapConnection-37118] davmail  - > PAPH7 OK [READ-WRITE] SELECT completed
2019-01-18 09:26:47,539 DEBUG [ImapConnection-37118] davmail  - < PAPH8 UID SEARCH ALL
2019-01-18 09:26:47,593 DEBUG [ImapConnection-37118] davmail.exchange.ExchangeSession  - Folder INBOX - Search items current count: 4 fetchCount: 500 highest uid: 14 lowest uid: 10
2019-01-18 09:26:47,593 DEBUG [ImapConnection-37118] davmail.exchange.ExchangeSession  - Message IMAP uid: 10 uid: AAAAAMJFXiqcxYZLn8LeyLhsBJ8BAOkEKOL0zPtKlch7lA7QutsAAAAADSgAAA== ItemId: AAMkAGZlODU3MGNkLThhNjEtNDZmOS05MjIxLTFkMjZjYzE5YjVhZgBGAAAAAADCRV4qnMWGS5/C3si4bASfBwDpBCji9Mz7SpXIe5QO0LrbAAAAAAEMAADpBCji9Mz7SpXIe5QO0LrbAAAAAA0oAAA= ChangeKey: CQAAABYAAADpBCji9Mz7SpXIe5QO0LrbAAAAABDQ
2019-01-18 09:26:47,593 DEBUG [ImapConnection-37118] davmail.exchange.ExchangeSession  - Message IMAP uid: 11 uid: AAAAAMJFXiqcxYZLn8LeyLhsBJ8BAOkEKOL0zPtKlch7lA7QutsAAAAADSkAAA== ItemId: AAMkAGZlODU3MGNkLThhNjEtNDZmOS05MjIxLTFkMjZjYzE5YjVhZgBGAAAAAADCRV4qnMWGS5/C3si4bASfBwDpBCji9Mz7SpXIe5QO0LrbAAAAAAEMAADpBCji9Mz7SpXIe5QO0LrbAAAAAA0pAAA= ChangeKey: CQAAABYAAADpBCji9Mz7SpXIe5QO0LrbAAAAABDR
2019-01-18 09:26:47,594 DEBUG [ImapConnection-37118] davmail.exchange.ExchangeSession  - Message IMAP uid: 13 uid: AAAAAMJFXiqcxYZLn8LeyLhsBJ8BAOkEKOL0zPtKlch7lA7QutsAAAAADSsAAA== ItemId: AAMkAGZlODU3MGNkLThhNjEtNDZmOS05MjIxLTFkMjZjYzE5YjVhZgBGAAAAAADCRV4qnMWGS5/C3si4bASfBwDpBCji9Mz7SpXIe5QO0LrbAAAAAAEMAADpBCji9Mz7SpXIe5QO0LrbAAAAAA0rAAA= ChangeKey: CQAAABYAAADpBCji9Mz7SpXIe5QO0LrbAAAAABDV
2019-01-18 09:26:47,594 DEBUG [ImapConnection-37118] davmail.exchange.ExchangeSession  - Message IMAP uid: 14 uid: AAAAAMJFXiqcxYZLn8LeyLhsBJ8BAOkEKOL0zPtKlch7lA7QutsAAAAADSwAAA== ItemId: AAMkAGZlODU3MGNkLThhNjEtNDZmOS05MjIxLTFkMjZjYzE5YjVhZgBGAAAAAADCRV4qnMWGS5/C3si4bASfBwDpBCji9Mz7SpXIe5QO0LrbAAAAAAEMAADpBCji9Mz7SpXIe5QO0LrbAAAAAA0sAAA= ChangeKey: CQAAABYAAADpBCji9Mz7SpXIe5QO0LrbAAAAABDW
2019-01-18 09:26:47,595 DEBUG [ImapConnection-37118] davmail  - > * SEARCH 10 11 13 14
2019-01-18 09:26:47,595 DEBUG [ImapConnection-37118] davmail  - > PAPH8 OK SEARCH completed
2019-01-18 09:26:47,639 DEBUG [ImapConnection-37118] davmail  - < PAPH9 UID FETCH 10 (RFC822)
2019-01-18 09:26:47,651 DEBUG [ImapConnection-37118] davmail.imap.ImapConnection  - * 1 FETCH (UID 10
2019-01-18 09:26:47,669 DEBUG [ImapConnection-37118] davmail.exchange.ExchangeSession  - Downloaded full message content for IMAP UID 10 (12048 bytes)
2019-01-18 09:26:47,670 DEBUG [ImapConnection-37118] davmail  - >  BODY[null] {12048}
2019-01-18 09:26:47,670 DEBUG [ImapConnection-37118] davmail  - > )
2019-01-18 09:26:47,670 DEBUG [ImapConnection-37118] davmail  - > PAPH9 OK UID FETCH completed
2019-01-18 09:26:47,770 INFO  [ImapConnection-37118] davmail.connection  - DISCONNECT - 127.0.0.1:37118

**Comment by seanthegeek** [comment]
Whoops! I made the change, but didn't actually push it to Github. Sorry about that. Please try again.

**Comment by stefankahn** [comment]
No worries.
Upgraded and re run.

   DEBUG:__init__.py:905:Connecting to IMAP over plain text
   DEBUG:__init__.py:976:Found 4 messages in IMAP folder INBOX
   DEBUG:__init__.py:982:Processing message 1 of 4: UID 10
Traceback (most recent call last):
  File "/usr/local/bin/parsedmarc", line 9, in <module>
    load_entry_point('parsedmarc==5.2.1', 'console_scripts', 'parsedmarc')()
  File "/usr/local/lib/python3.5/dist-packages/parsedmarc/cli.py", line 343, in _main
    strip_attachment_payloads=sa
  File "/usr/local/lib/python3.5/dist-packages/parsedmarc/__init__.py", line 989, in get_dmarc_reports_from_inbox
    ensure_ascii=True))
  File "/usr/lib/python3.5/json/__init__.py", line 237, in dumps
    **kw).encode(obj)
  File "/usr/lib/python3.5/json/encoder.py", line 200, in encode
    chunks = list(chunks)
  File "/usr/lib/python3.5/json/encoder.py", line 429, in _iterencode
    yield from _iterencode_dict(o, _current_indent_level)
  File "/usr/lib/python3.5/json/encoder.py", line 403, in _iterencode_dict
    yield from chunks
  File "/usr/lib/python3.5/json/encoder.py", line 375, in _iterencode_dict
    raise TypeError("key " + repr(key) + " is not a string")
TypeError: key b'BODY[NULL]' is not a string

**Comment by seanthegeek** [comment]
Not the result I was expecting. I just pushed another change. Try upgrading and running again.

**Comment by stefankahn** [comment]
Upgraded and run.

   DEBUG:__init__.py:906:Connecting to IMAP over plain text
   DEBUG:__init__.py:977:Found 4 messages in IMAP folder INBOX
   DEBUG:__init__.py:983:Processing message 1 of 4: UID 10
Traceback (most recent call last):
  File "/usr/local/bin/parsedmarc", line 9, in <module>
    load_entry_point('parsedmarc==5.2.1', 'console_scripts', 'parsedmarc')()
  File "/usr/local/lib/python3.5/dist-packages/parsedmarc/cli.py", line 343, in _main
    strip_attachment_payloads=sa
  File "/usr/local/lib/python3.5/dist-packages/parsedmarc/__init__.py", line 989, in get_dmarc_reports_from_inbox
    PrettyPrinter.pprint(raw_msg)
TypeError: pprint() missing 1 required positional argument: 'object'

**Comment by mfritts** [comment]
I've been running parsedmarc as a service for several months. It's working well with the kibana dashboard. But I'm having a problem now that I'm getting about 200 reports a day. Is there a way to run in batches of 10 messages or so? My mail server is timing out and closing the connection before the reports can be archived. If I could run in batches or at least archive the messages that it's already processed instead of starting over when it sees more messages have arrived in the inbox before the last run is finished I think I could get around my timeout issue. 
But I'm also looking into other ideas for dealing with this, like running on more than one inbox and process, to reduce the load. Maybe you have some suggestions if no changes to the script?
Great tool, though! I've been pretty happy so far which is why I increased the work that it was doing so much. lol.

**Comment by seanthegeek** [comment]
Sony for the delay. Try the latest dev release from git again. That debug output should tell me what I need to know to fix this.

**Comment by stefankahn** [comment]
Apologies we had a long weekend here.
Updated and rerun

   DEBUG:__init__.py:905:Connecting to IMAP over plain text
   DEBUG:__init__.py:828:IMAP server supports: ['IMAP4REV1', 'AUTH=LOGIN', 'IDLE', 'MOVE', 'SPECIAL-USE']
   DEBUG:__init__.py:976:Found 22 messages in IMAP folder INBOX
   DEBUG:__init__.py:982:Processing message 1 of 22: UID 10
Traceback (most recent call last):
  File "/usr/local/bin/parsedmarc", line 9, in <module>
    load_entry_point('parsedmarc==5.2.1', 'console_scripts', 'parsedmarc')()
  File "/usr/local/lib/python3.5/dist-packages/parsedmarc/cli.py", line 343, in _main
    strip_attachment_payloads=sa
  File "/usr/local/lib/python3.5/dist-packages/parsedmarc/__init__.py", line 989, in get_dmarc_reports_from_inbox
    raw_msg = raw_msg[message_uid][b"RFC822"]
KeyError: b'RFC822'

**Comment by seanthegeek** [comment]
I finally had time to reproduce this myself. Fixed in parsedmarc 5.3.0! :)
